### ISSUE TITLE
Usage of ProductType

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#4950]
We use `ProductType<A,B>::type` to model what happens when one multiplies an `A` and a `B` object. Conceptually, it should be something like
```
  decltype (declval<A>() * declval<B>())
```
and indeed, that is how it is defined.

Whether or not `A` and `B` are const or reference qualified should really not matter here, because it should not affect the *result* of taking the product. Yet, we strip qualifiers from `A` and `B` in many places where it is used. For example, #4947 make this kind of change that doesn't appear obvious to me (but apparently helps):
```
-      typedef typename ProductType<typename std::remove_cv<Number>::type, typename Scalar<dim,spacedim>::value_type>::type value_type;
+      typedef typename ProductType<typename std::decay<Number>::type, typename Scalar<dim,spacedim>::value_type>::type value_type;
```
I don't understand why we need to decay the type, but neither do I understand why we needed to remove CV qualifiers before.

A pragmatic approach would be to move the `std::decay` into the definition of `ProductType::value` so it doesn't have to be repeated as many times. At the same time, I'd first like to understand why we need either `std::remove_cv` or `std::decay` to begin with.

@jppelteret -- can you comment on what kind of errors you got before your change in #4947 ?

**Comment by jppelteret** [comment]
I wish that I'd kept the error messages, so I'll have to reproduce one later to demonstrate what can go wrong here. Basically, the default `ProductType<A,B>::type` is as you stated, but to the best of my recollection the trouble comes in with specialisations. For example, we specialise `ProductType<Sacado::Fad::DFad,double>` but not `ProductType<const Sacado::Fad::DFad,double>`. So internally somewhere (I think from `ArrayView`) we pass a `const NumberType` that propagates to `ProductType` and the (my?) compiler doesn't know what to do with it.

Maybe it'll be a bit more clear if I produce a small test case. I'll post it here once I've done so.

**Comment by bangerth** [comment]
Ah, I see. That makes sense. 

I think we could work around this by using something along the following lines:
```
namespace internal {
  template <typename T, typename U>
  struct ProductTypeImpl
  {
    typedef decltype(std::declval<T>() * std::declval<U>()) type;
  };
}

template <typename T, typename U>
struct ProductType
{
  typedef internal::ProductTypeImpl<std::decay<T>::type, std::decay<U>::type>::type type;
};
```
Then we would only need to specialize `internal::ProductTypeImpl` for which we know that it will only ever get stripped types. This way we wouldn't have to pass stripped types all over the place but do it in one place.

But I'd still be curious to hear about the difference between `std::remove_cv` and `std::decay` if you could dig up an error message that shows the difference.

**Comment by jppelteret** [comment]
Now you're having me question whether I absolutely needed `std::decay` or if I just wanted to hit the problem with the biggest bat.

I like your suggestion, and would be happy to put together a patch that implements it.

**Comment by masterleinad** [comment]
Closed by #4951.
